File: python/tests/docs/test_subgraph_attr.py
from sp_svg_diagram import SVGDiagram

from .utils import compare_svg


def test_inherited():
    diagram = SVGDiagram()
    subgraph_inner = diagram.add_subgraph("subgraph-inner")
    subgraph_outer = diagram.add_subgraph("subgraph-outer")
    subgraph_outer.add_subgraph(subgraph_inner)

    node1 = diagram.add_node("A")
    node1.set_label("A")
    node1.set_center(0, 0)
    node2 = diagram.add_node("B")
    node2.set_label("B")
    node2.set_center(150, 0)
    edge1 = diagram.add_edge("A", "B")
    edge1.set_arrow_head()
    subgraph_inner.add_node(node1)
    subgraph_inner.add_node(node2)
    subgraph_inner.add_edge(edge1)
    subgraph_inner.default_node_attributes().set_pen_width(2)
    subgraph_inner.default_edge_attributes().set_pen_width(2)

    node3 = diagram.add_node("C")
    node3.set_label("C")
    node3.set_center(0, 60)
    node4 = diagram.add_node("D")
    node4.set_label("D")
    node4.set_center(150, 60)
    edge2 = diagram.add_edge("C", "D")
    edge2.set_arrow_head()
    edge3 = diagram.add_edge("B", "C")
    edge3.set_arrow_head()
    subgraph_outer.add_node(node3)
    subgraph_outer.add_node(node4)
    subgraph_outer.add_edge(edge2)
    subgraph_outer.add_edge(edge3)
    subgraph_outer.default_node_attributes().set_fill_color("lightgray")

    svg = diagram.render()
    compare_svg("subgraph_attr", "inherited", svg)

File: tests/docs/test_subgraph_attr.cpp
#include "svg_diagram.h"

#include <cmath>
#include <numbers>
#include <format>
#include <vector>
#include <gtest/gtest.h>
using namespace svg_diagram;

static const auto OUTPUT_DIR = std::string("../docs/source/_static/subgraph_attr/");

TEST(TestDocsSubgraphAttributes, Inherited) {
    SVGDiagram diagram;
    auto subgraphInner = diagram.addSubgraph("subgraph-inner");
    const auto subgraphOuter = diagram.addSubgraph("subgraph-outer");
    subgraphOuter->addSubgraph(subgraphInner);

    auto node1 = diagram.addNode("A");
    node1->setLabel("A");
    node1->setCenter(0, 0);
    auto node2 = diagram.addNode("B");
    node2->setLabel("B");
    node2->setCenter(150, 0);
    auto edge1 = diagram.addEdge("A", "B");
    edge1->setArrowHead();
    subgraphInner->addNode(node1);
    subgraphInner->addNode(node2);
    subgraphInner->addEdge(edge1);
    subgraphInner->defaultNodeAttributes().setPenWidth(2);
    subgraphInner->defaultEdgeAttributes().setPenWidth(2);

    auto node3 = diagram.addNode("C");
    node3->setLabel("C");
    node3->setCenter(0, 60);
    auto node4 = diagram.addNode("D");
    node4->setLabel("D");
    node4->setCenter(150, 60);
    auto edge2 = diagram.addEdge("C", "D");
    edge2->setArrowHead();
    auto edge3 = diagram.addEdge("B", "C");
    edge3->setArrowHead();
    subgraphOuter->addNode(node3);
    subgraphOuter->addNode(node4);
    subgraphOuter->addEdge(edge2);
    subgraphOuter->addEdge(edge3);
    subgraphOuter->defaultNodeAttributes().setFillColor("lightgray");

    diagram.render(OUTPUT_DIR + "inherited.svg");
}

TEST(TestDocsSubgraphAttributes, Draw) {
    SVGDiagram diagram;
    auto subgraphInner = std::make_shared<SVGGraph>("subgraph-inner");
    subgraphInner->setLabel("Inner");
    subgraphInner->setColor("black");
    subgraphInner->setFillColor("lightgreen");
    const auto subgraphOuter = diagram.addSubgraph("subgraph-outer");
    subgraphOuter->addSubgraph(subgraphInner);
    subgraphOuter->setLabel("Outer");
    subgraphOuter->setColor("black");
    subgraphOuter->setFillColor("papayawhip");
    auto node1 = std::make_shared<SVGNode>("A");
    node1->setLabel("A");
    node1->setCenter(0, 0);
    auto node2 = std::make_shared<SVGNode>("B");
    node2->setLabel("B");
    node2->setCenter(150, 0);
    const auto edge = diagram.addEdge("A", "B");
    edge->setArrowHead();
    subgraphInner->addNode(node1);
    subgraphOuter->addNode(node2);
    diagram.render(OUTPUT_DIR + "draw.svg");
}